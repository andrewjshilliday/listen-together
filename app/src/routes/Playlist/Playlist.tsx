import React, { useState, useEffect, createRef, useLayoutEffect, useRef } from 'react';
import moment from 'moment';
import { MusicKitService, MusicKitApiService } from '../../services';
import { Loading, TrackList, MediaItemCardCarousel } from '../../components/common';
import styled from 'styled-components';

const Playlist: React.FC = (props: any) => {
  const id = props.match.params.id;
  const [loading, setLoading] = useState(true);
  const [playlist, setPlaylist] = useState({} as MusicKit.MediaItem);
  const [featuredArtists, setFeaturedArtists] = useState([] as MusicKit.MediaItem[]);
  const [duration, setDuration] = useState('');
  const playlistRef = useRef<MusicKit.MediaItem>(playlist);
  const notesRef = createRef<HTMLDivElement>();

  useEffect(() => {
    const getPlaylist = async () => {
      setLoading(true);
      setPlaylist(await MusicKitApiService.Playlist(id));
      setLoading(false);
    };
    getPlaylist();
  }, [id]);

  useEffect(() => {
    if (playlist.relationships) {
      let durationInMillis
      let hours = 0, minutes = 0;
      let durationString = '';

      durationInMillis = moment.duration(playlist.relationships.tracks.data.map((i: MusicKit.MediaItem) => i.attributes.durationInMillis)
        .reduce((a: number, b: number) => a + b, 0));
      hours = Math.floor(durationInMillis.asHours());
      durationInMillis = moment.duration(durationInMillis.asMilliseconds() - (hours * 60 * 60 * 1000));
      minutes = Math.round(durationInMillis.asMinutes());

      if (hours) {
        durationString += `${hours} Hour`;
        if (hours > 1) { durationString += 's' }
        durationString += ' '
      }
      durationString += `${minutes} Minute`;
      if (minutes > 1) { durationString += 's' }

      setDuration(durationString);
    }

    const getTrackRelationships = async () => {
      if (!playlist.relationships) { return }
      const tracks = await MusicKitApiService.Songs(playlist.relationships.tracks.data.map((i: MusicKit.MediaItem) => i.id));
      const artistIds = tracks.map((r: MusicKit.MediaItem) => r.relationships.artists.data[0].id)
        .filter((v: MusicKit.MediaItem, i: number, a: MusicKit.MediaItem[]) => a.indexOf(v) === i);
      setFeaturedArtists(await MusicKitApiService.Artists(artistIds));
    }
    getTrackRelationships();
  }, [playlist]);

  useLayoutEffect(() => {
    setEditorialNotesStyle();
    window.addEventListener('resize', setEditorialNotesStyle);
    return () => window.removeEventListener('resize', setEditorialNotesStyle);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [notesRef]);
  
  const setEditorialNotesStyle = () => {
    if (!playlistRef.current?.attributes?.editorialNotes || !notesRef.current) {
      return;
    }

    const height = window.innerHeight;
    const notesOffsetTop = notesRef.current.getBoundingClientRect().top;
    notesRef.current.style.maxHeight = `${height - notesOffsetTop - 110}px`;
  };

  if (loading) return (<Loading />);
  
  return (
    <PlaylistContainer>
      <PlaylistTracksContainer>
        <SidebarContainer>
          <StickySidebar>
            <img src={MusicKitService.FormatArtwork(playlist.attributes.artwork, 500)} className="img-fluid rounded" alt={playlist.attributes.name} />
            <p>{playlist.relationships.tracks.data.length} Songs, {duration}</p>
            {
              playlist.attributes.editorialNotes &&
              <>
                <hr />
                <EditorialNotes ref={notesRef}>
                {
                  playlist.attributes.editorialNotes.standard ?
                  <span dangerouslySetInnerHTML={{ __html: `${playlist.attributes.editorialNotes.standard}` }}></span> :
                  <span dangerouslySetInnerHTML={{ __html: `${playlist.attributes.editorialNotes.short}`}}></span>
                }
                </EditorialNotes>
              </>
            }
          </StickySidebar>
        </SidebarContainer>
        <TrackListContainer>
          <TrackListHeader>
            <h1 className="text-truncate">{playlist.attributes.name}</h1>
            <CuratorLastModified>
              <h2 className="text-truncate">{playlist.attributes.curatorName}</h2>
              <h2 title={moment.utc(playlist.attributes.lastModifiedDate).format('DD/MM/YYYY HH:mm:ss')}>&nbsp;| Last Modified: {moment.utc(playlist.attributes.lastModifiedDate).fromNow()}</h2>
            </CuratorLastModified>
          </TrackListHeader>
          <TrackList tracks={playlist.relationships.tracks.data} showArtist={true} showAlbum={true}></TrackList>
        </TrackListContainer>
      </PlaylistTracksContainer>
      {
        featuredArtists.length &&
        <FeaturedArtistsContainer>
          <MediaItemCardCarousel items={featuredArtists} title='Featured Artists'></MediaItemCardCarousel>
        </FeaturedArtistsContainer>
      }
    </PlaylistContainer>
  );
}

export default Playlist;


const PlaylistContainer = styled.div``;
const PlaylistTracksContainer = styled.div`
  display: flex;
`;
const SidebarContainer = styled.div`
  position: sticky;
  top: 0;
  width: 30%;
  min-width: 250px;
  max-width: 400px;
  padding: 10px 15px;
`;
const StickySidebar = styled.div`
  position: sticky;
  top: 10px;
`;
const EditorialNotes = styled.div`
  overflow: auto;
  font-size: 10pt;
`;
const TrackListContainer = styled.div`
  width: 70%;
  padding-left: 15px;
`;
const TrackListHeader = styled.div`
  position: sticky;
  top: 0;
  padding-bottom: 10px;
  background: white;
`;
const CuratorLastModified = styled.div`
  display: flex;
`;
const FeaturedArtistsContainer = styled.div`
  margin: 0 10px;
`;
