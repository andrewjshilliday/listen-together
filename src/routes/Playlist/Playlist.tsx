import React, { useState, useEffect, createRef } from 'react';
import { MKPlaylist } from '../../services/MusicKitApi';
import { FormatArtwork } from '../../services/MusicKit';
import { TrackList } from '../../components/common';
import CircularProgress from '@material-ui/core/CircularProgress';
import './Playlist.scss';

const Playlist: React.FC = (props: any) => {
  const [playlist, setPlaylist] = useState({} as any);
  const [loading, setLoading] = useState(true);
  const id = props.match.params.id;
  const notesRef = createRef<HTMLDivElement>();

  useEffect(() => {
    const getPlaylist = async () => {
      setLoading(true);
      setPlaylist(await MKPlaylist(id));
      setLoading(false);
    };
    getPlaylist();
  }, [id]);

  useEffect(() => {
    setEditorialNotesStyle();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [notesRef]);

  window.onresize = () => {
    setEditorialNotesStyle();
  };
  
  const setEditorialNotesStyle = () => {
    if (!playlist.attributes || !playlist.attributes.editorialNotes) {
      return;
    }

    if (notesRef.current) {
      const height = window.innerHeight;
      const notesOffsetTop = notesRef.current.getBoundingClientRect().top;
      notesRef.current.style.maxHeight = `${height - notesOffsetTop - 110}px`;
    }
  };

  if (loading) return (<div className="loading"><CircularProgress /></div>);
  
  return (
    <div className="playlist view">
      <div className="sidebar sticky">
        <img src={FormatArtwork(playlist.attributes.artwork, 500)} className="img-fluid rounded" alt={playlist.attributes.name} />
        {
          playlist.attributes.editorialNotes &&
          <>
            <hr />
            <div ref={notesRef} className="editorial-notes">
            {
              playlist.attributes.editorialNotes.standard ?
              <span dangerouslySetInnerHTML={{ __html: `${playlist.attributes.editorialNotes.standard}` }}></span> :
              <span dangerouslySetInnerHTML={{ __html: `${playlist.attributes.editorialNotes.short}`}}></span>
            }
            </div>
          </>
        }
      </div>
      <div className="track-list-container">
        <div className="track-list-header">
          <h1 className="text-truncate">{playlist.attributes.name}</h1>
          <h2 className="text-truncate">{playlist.attributes.artistName} {playlist.attributes.curatorName} | {playlist.attributes.lastModified}</h2>
        </div>
        <TrackList tracks={playlist.relationships.tracks.data} displayArtwork={false}></TrackList>
      </div>
    </div>
  );
}

export default Playlist;
