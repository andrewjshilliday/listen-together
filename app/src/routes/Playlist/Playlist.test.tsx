import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import Playlist from './Playlist';

jest.mock('../../services/MusicKit');
jest.mock('../../services/MusicKitApi', () => ({
  MusicKitApiService: require('../../__mocks__/services/MusicKitApi/MusicKitApi.mock').default
}));

jest.useFakeTimers();

it('renders correctly', async () => {  
  const tree = renderer.create(
    <MemoryRouter initialEntries={[ '/playlist/pl.a5ef67f3dde74a568e583ac9f0406d50' ]}>
      <Route path="/playlist/:id" component={Playlist} />
    </MemoryRouter>);

  jest.runTimersToTime(1);

  await act(async () => {
    await Promise.resolve();
  });

  expect(tree.toJSON()).toMatchSnapshot();
});
