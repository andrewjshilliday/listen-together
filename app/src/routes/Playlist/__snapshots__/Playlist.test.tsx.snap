// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-fzoiQi jkvegY"
>
  <div
    className="sc-fzqARJ lnyOvm"
  >
    <div
      className="sc-fzqNqU fuEiiX"
    >
      <div
        className="sc-fzoyTs iVKUqR"
      >
        <img
          alt="Gary Clark Jr. Essentials"
          className="img-fluid rounded"
        />
        <p>
          17
           Songs, 
          1 Hour 25 Minutes
        </p>
      </div>
    </div>
    <div
      className="sc-fzoXWK fniOzM"
    >
      <div
        className="sc-fzpmMD eWMA-di"
      >
        <h1
          className="text-truncate"
        >
          Gary Clark Jr. Essentials
        </h1>
        <div
          className="sc-fznxKY jzQXIe"
        >
          <h2
            className="text-truncate"
          >
            Apple Music Blues
          </h2>
          <h2
            title="08/07/2019 23:38:54"
          >
            Â | Last Modified: 
            a year ago
          </h2>
        </div>
      </div>
      <div
        className="sc-AxheI fniENP"
      >
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Got to Get Up
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            This Land
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            2:37
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Bright Lights
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            5:24
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Things Are Changin' (Live) [Solo Acoustic]
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            The Bright Lights - EP
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:31
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Grinder
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            The Story of Sonny Boy Slim
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            3:22
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            This Land
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            This Land
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            5:41
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Don't Owe You a Thang
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            The Bright Lights - EP
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            3:35
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Numb
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            5:26
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Our Love (Live)
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Live North America 2016
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            7:50
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            The Healing
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            The Story of Sonny Boy Slim
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:48
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Catfish Blues (Live)
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Gary Clark Jr. Live
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            8:00
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            When My Train Pulls In
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            7:45
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Ain't Messin 'Round
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:09
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Please Come Home
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:16
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Church
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            The Story of Sonny Boy Slim
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            5:10
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Take Me Down
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Take Me Down - Single
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:47
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Travis County
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Blak and Blu (Deluxe Version)
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            3:38
          </span>
        </div>
        <div
          className="sc-Axmtr cffNzl"
        >
          <div
            className="sc-AxmLO eHiwgP"
            onClick={[Function]}
          >
            <i
              className="fas fa-plus"
            />
          </div>
          <span
            className="sc-fzozJi hQGPdZ text-truncate"
          >
            Next Door Neighbor Blues (Live)
          </span>
          <span
            className="sc-fzoLsD gfgwZO text-truncate"
          >
            Gary Clark Jr.
          </span>
          <span
            className="sc-fzpans iAAgBG track-album text-truncate"
          >
            Gary Clark Jr. Live
          </span>
          <span
            className="sc-fzplWN eXRBjQ track-duration"
          >
            4:06
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-fznMAR MCjQk"
  >
    <div
      className="sc-AxirZ fJUMyY"
    >
      <h2>
        Featured Artists
      </h2>
      <div
        className="sc-AxiKw enaPwj"
      >
        <div
          className="sc-AxhCb iEIFwb"
          onClick={[Function]}
        />
        <div
          className="sc-AxgMl eFueIV"
        >
          <div
            className="sc-AxjAm dYpDUQ"
          >
            <a
              href="/artist/266463936"
              onClick={[Function]}
            >
              <img
                alt="Gary Clark Jr."
                className="img-fluid rounded"
                src="placeholder.jpeg"
              />
              <span
                className="text-truncate"
              >
                Gary Clark Jr.
              </span>
            </a>
          </div>
        </div>
        <div
          className="sc-AxhCb sc-AxhUy ixXRnl"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
